 

HystrixInvocationHandler那儿去，断点本来就有

 

就跑起来服务的环境，然后发送请求，来继续断点调试，看看人家的HystrixCommand的execute()方法，去执行一个请求的入口

 

return HystrixCommand.class.isAssignableFrom(method.getReturnType());

 

method.getReturnType()：你要调用的接口方法的返回类型

 

  if (isReturnsHystrixCommand(method)) {

   return hystrixCommand;

  } else if (isReturnsObservable(method)) {

   // Create a cold Observable

   return hystrixCommand.toObservable();

  } else if (isReturnsSingle(method)) {

   // Create a cold Observable as a Single

   return hystrixCommand.toObservable().toSingle();

  } else if (isReturnsCompletable(method)) {

   return hystrixCommand.toObservable().toCompletable();

}

 

isReturnsHystrixCommand(method)：判断一下method的返回类型，是否是HystrixCommand类型

 

isReturnsObservable(method)：判断一下method的返回类型，是否是Observable类型

 

他会判断一下，你的接口方法

 

HystrixCommand sayHello(Long id, String name, Integer age)

 

sayHello()方法的返回类型是HystrixCommand的话，那么在这里，针对这个方法的调用，构造好了HystrixCommand之后，就不会去执行这个command，而是将这个command作为你的sayHello()方法的返回值，返回给你

 

让你来决定，如何使用这个command

 

大家通过源码的阅读，源码多有意思，可以非常深入、透彻的理解feign和hystirx整合的时候，应该怎么用

 

execute()方法 -> 

 

return queue().get();

 

queue()方法，感觉这个queue()方法是什么东西啊？

 

从这个queue()方法开始，我们就要研究netflix hystrix的源码了

 

queue()方法，是用来异步执行command业务逻辑的，他会将command扔到线程池里去执行，但是这个方法是不会等待这个线程执行完毕command的，他会拿到一个Future对象，通过Future对象去获取command执行完毕的响应结果

 

​    /*

​     \* The Future returned by Observable.toBlocking().toFuture() does not implement the

​     \* interruption of the execution thread when the "mayInterrupt" flag of Future.cancel(boolean) is set to true;

​     \* thus, to comply with the contract of Future, we must wrap around it.

​     */

​    final Future<R> delegate = toObservable().toBlocking().toFuture();

 

toObservable()：Observable对象

 

toObservable().toBlocking().toFuture()：此时这个command已经被扔到了线程池里去执行了，获取了一个线程池里执行线程对应的一个Future对象

 

这里你获取到的这个Future对象，是不具备，因为一些异常的原因，中断这个线程执行的能力的，比如超时、异常，你没办法在异常情况下，终止future对应的线程的执行，所以说要对这里返回的delegate future进行包装

 

​    final Future<R> f = new Future<R>() {

 

​      @Override

​      public boolean cancel(boolean mayInterruptIfRunning) {

​        if (delegate.isCancelled()) {

​          return false;

​        }

 

​        if (HystrixCommand.this.getProperties().executionIsolationThreadInterruptOnFutureCancel().get()) {

​          /*

​           \* The only valid transition here is false -> true. If there are two futures, say f1 and f2, created by this command

​           \* (which is super-weird, but has never been prohibited), and calls to f1.cancel(true) and to f2.cancel(false) are

​           \* issued by different threads, it's unclear about what value would be used by the time mayInterruptOnCancel is checked.

​           \* The most consistent way to deal with this scenario is to say that if *any* cancellation is invoked with interruption,

​           \* than that interruption request cannot be taken back.

​           */

​          interruptOnFutureCancel.compareAndSet(false, mayInterruptIfRunning);

​    		}

 

​        final boolean res = delegate.cancel(interruptOnFutureCancel.get());

 

​        if (!isExecutionComplete() && interruptOnFutureCancel.get()) {

​          final Thread t = executionThread.get();

​          if (t != null && !t.equals(Thread.currentThread())) {

​            t.interrupt();

​          }

​        }

 

​        return res;

​			}

 

​      @Override

​      public boolean isCancelled() {

​        return delegate.isCancelled();

​			}

 

​      @Override

​      public boolean isDone() {

​        return delegate.isDone();

​			}

 

​      @Override

​      public R get() throws InterruptedException, ExecutionException {

​        return delegate.get();

​      }

 

​      @Override

​      public R get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {

​        return delegate.get(timeout, unit);

​      }

​    	

​    };

 

所以说这个新创建的future对象，其实包装了上面的原生的delegate future对象，这里的cancel()方法，就是说，可以支持将delegate future对象对应的thread线程interrupt中止掉，封装了一些isCanceled()、isDone()、get()等future相关的代码

 

他搞两个future，最主要的是用第二个future，去包装第一个future，第一个future，原生的情况下，是不支持cancel()方法中止自己对应的那个在执行的线程的，所以搞第二个future包装一下，可以支持cancel()中止线程

 

hystrix command对应着一个服务接口，属于一个服务，服务就是对应@FeignClient里的服务名称，ServiceA，一个服务是一个groupKey，对应一个线程池，一定是要基于服务的线程池来进行执行的

 

​    if (f.isDone()) {

​      try {

​        f.get();

​        return f;

​      } catch (Exception e) {

​        Throwable t = decomposeException(e);

​        if (t instanceof HystrixBadRequestException) {

​          return f;

​        } else if (t instanceof HystrixRuntimeException) {

​          HystrixRuntimeException hre = (HystrixRuntimeException) t;

​          switch (hre.getFailureType()) {

​					case COMMAND_EXCEPTION:

​					case TIMEOUT:

​						// we don't throw these types from queue() only from queue().get() as they are execution errors

​						return f;

​					default:

​						// these are errors we throw from queue() as they as rejection type errors

​						throw hre;

​					}

​        } else {

​          throw Exceptions.sneakyThrow(t);

​        }

​      }

​    }

 

f.isDone()，就是通过future判断对应的那个线程，是否完成了command的执行

 

f.get()，这里会阻塞住，卡住，尝试去通过future获取对应的thread对command执行后返回的结果，这里会卡住

 

如果这行代码，执行完毕了，那么就说明可以获取到thread执行command返回的结果

 

此时就会直接返回future

 

  public R execute() {

​    try {

​      return queue().get();

​    } catch (Exception e) {

​      throw Exceptions.sneakyThrow(decomposeException(e));

​    }

}

 

queue()了以后，就会通过线程池去执行command，然后在queue()方法中，会等待线程执行结束，如果线程执行结束了，就会返回future；即使执行失败了，也会根据情况，返回future，要不就是抛异常

 

在execute()方法中，一定会拿到future，然后直接调用future.get()方法，尝试去获取futrue对应的thread执行command的一个结果

 

execute()方法和queue()方法，主要的执行逻辑和流程

 

下一讲开始，我们重点要研究的是哪儿？

 

toObservable().toBlocking().toFuture()

 

这行代码，才是重中之重，这行代码底层实现了hystirx几乎所有的核心逻辑，请求缓存、熔断、队列+线程池、线程异步执行、超时检测、异常处理、异常统计、熔断开关